### PR TITLE
Feature/okhttp support

### PR DESCRIPTION
**Pull request by falvick**
Basic functionality without unnecessary code and libs

### CODE REVIEW

File: slack-api/build.gradle [hunk 1]
```diff
@@ -31,16 +31,15 @@ android {
 }
 
 dependencies {
-    implementation "com.fasterxml.jackson.module:jackson-module-kotlin:2.9.0"
```
Comment (m0d) [inline slack-api/build.gradle:34]: ❤️ ❤️ ❤️ ❤️ ❤️ ❤️ ❤️ ❤️ ❤️ ❤️ ❤️ ❤️ ❤️

File: slack-api/build.gradle [hunk 2]
```diff
@@ -31,16 +31,15 @@ android {
...
-    implementation "cz.msebera.android:httpclient:4.4.1.2"
```
Comment (m0d) [inline slack-api/build.gradle:35]: ❤️ ❤️ ❤️ ❤️ ❤️ ❤️ ❤️ ❤️ ❤️ ❤️ ❤️ ❤️ ❤️

File: slack-api/build.gradle [hunk 3]
```diff
@@ -31,16 +31,15 @@ android {
...
-    implementation "com.fasterxml.jackson.core:jackson-databind:2.9.0"
```
Comment (m0d) [inline slack-api/build.gradle:36]: ❤️ ❤️ ❤️ ❤️ ❤️ ❤️ ❤️ ❤️ ❤️ ❤️ ❤️ ❤️ ❤️

File: slack-api/src/main/java/pl/hellsoft/slack/wrapper/rtm/SlackRealTimeMessagingClient.kt [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+                if (eventListeners != null && !eventListeners.isEmpty()) {
+                    for (listener in eventListeners) {
+                        listener.onMessage(text)
+                    }
+                }
+            }
+        }
+
+        override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
+            if (!closeListeners.isEmpty()) {
```
Comment (m0d) [inline slack-api/src/main/java/pl/hellsoft/slack/wrapper/rtm/SlackRealTimeMessagingClient.kt]: body can be simplified
```
            closeListeners.forEach{
                it.onClose()
            }
```

File: slack-api/src/main/java/pl/hellsoft/slack/wrapper/rtm/SlackRealTimeMessagingClient.kt [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+                    listener.onClose()
+                }
+            }
+        }
+
+        override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
+            if (t !is Exception)
+                return
+
+            if (!failureListeners.isEmpty()) {
```
Comment (m0d) [inline slack-api/src/main/java/pl/hellsoft/slack/wrapper/rtm/SlackRealTimeMessagingClient.kt]: can be simplified

`                failureListeners.forEach{
                    it.onFailure(RuntimeException(t))
                }`

File: slack-api/src/main/java/pl/hellsoft/slack/wrapper/rtm/SlackRealTimeMessagingClient.kt [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+                    listener.onFailure(RuntimeException(t))
+                }
+            }
+        }
+    }
+
+    fun close() {
+        i { "Slack RTM closing..." }
+        try {
+            client?.dispatcher()?.executorService()?.shutdown()
```
Comment (m0d) [inline slack-api/src/main/java/pl/hellsoft/slack/wrapper/rtm/SlackRealTimeMessagingClient.kt]: can be simplified
```

            client?.dispatcher()?.run{
                executorService()?.shutdown()
                cancelAll()
            }
```

File: slack-api/src/main/java/pl/hellsoft/slack/wrapper/rtm/SlackRealTimeMessagingClient.kt [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+                    .writeTimeout(0, TimeUnit.MILLISECONDS)
+
+            if (proxyServerInfo != null) {
+                val proxy = Proxy(Proxy.Type.HTTP, InetSocketAddress(proxyServerInfo.host, proxyServerInfo.port))
+                clientBuilder.proxy(proxy)
+            }
+
+            val request = Request.Builder().url(webSocketUrl).build()
+            if (BuildConfig.DEBUG) {
+                val interceptor = HttpLoggingInterceptor()
```
Comment (m0d) [inline slack-api/src/main/java/pl/hellsoft/slack/wrapper/rtm/SlackRealTimeMessagingClient.kt]: interceptor for websocket is useless imho

File: slack-api/src/main/java/pl/hellsoft/slack/wrapper/rtm/SlackRealTimeMessagingClient.kt [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+                type = connect.type
+            } catch (e: Exception) {
+                e { "$e.message" }
+            }
+
+            if ("pong" != type) {
+                i { "Slack RTM message : $text" }
+            }
+
+            if (type != null) {
```
Comment (m0d) [inline slack-api/src/main/java/pl/hellsoft/slack/wrapper/rtm/SlackRealTimeMessagingClient.kt]: can be simplified
```

listeners[type]?.forEach{
                it.onMessage(text)
            }
```

File: slack-api/src/main/java/pl/hellsoft/slack/wrapper/rtm/SlackRealTimeMessagingClient.kt [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+            i { "Connected to Slack RTM (Real Time Messaging) server : " + webSocketUrl }
+        }
+
+        override fun onMessage(webSocket: WebSocket, text: String?) {
+            if (text == null)
+                return
+
+            var type: String? = null
+            try {
+                val connect = Gson().fromJson(text, RtmConnect::class.java)
```
Comment (m0d) [inline slack-api/src/main/java/pl/hellsoft/slack/wrapper/rtm/SlackRealTimeMessagingClient.kt]: maybe its worth to construct Gson once and hold the reference?

File: slack-api/src/main/java/pl/hellsoft/slack/wrapper/webapi/WebApiImpl.kt [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+import pl.hellsoft.slack.wrapper.SlackService
+
+/**
+ * @author Maciej Madetko
+ * @email maciej.madetko@ftlearning.com
+ * Nikkei FT Learning Limited
+ * @since 22/01/2018.
+ */
+class WebApiImpl(private val service: SlackService, private val token: String) : WrapperApiInterface {
+    override fun postMessage(channel: String, message: String) : Observable<PostMessageResponse> {
```
Comment (m0d) [inline slack-api/src/main/java/pl/hellsoft/slack/wrapper/webapi/WebApiImpl.kt:17]: rearrange function order
message related functions in one group

File: slack-api/src/main/java/pl/hellsoft/slack/wrapper/SlackApiWrapper.kt [hunk 11]
```diff
@@ -25,46 +34,80 @@ import pl.hellsoft.slack.wrapper.model.MySlack
  */
 
 open class SlackApiWrapper {
-    private var mToken : String? = null
-    private var mWebApiClient: SlackWebApiClient? = null
+    private var compositeDisposable = CompositeDisposable()
+    private lateinit var mToken : String
+    private val timeoutInSeconds = 6L
```
Comment (m0d) [inline slack-api/src/main/java/pl/hellsoft/slack/wrapper/SlackApiWrapper.kt]: use hungarian notation for privates

File: slack-api/src/main/java/pl/hellsoft/slack/wrapper/SlackService.kt [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+
+    @GET("${apiPath}/${SlackApiConstants.CHANNELS_HISTORY}")
+    fun getChannelHistoryEvents(
+            @Query("token") token: String,
+            @Query("channel") channel: String,
+            @Query("latest") latest: String?,
+            @Query("oldest") oldest: String?,
+            @Query("inclusive") inclusive: Boolean?,
+            @Query("count") count: Int,
+            @Query("unreads") unreads: Boolean?) : Observable<HistoryEvents>
```
Comment (m0d) [inline slack-api/src/main/java/pl/hellsoft/slack/wrapper/SlackService.kt]: if we uses XXXResponse Pattern so maybe it is worth to bbe consistent and change HistoryEvents to HistoryResponse and change getChannelHistoryEvents yo getChannelHistory

File: slack-api/src/main/java/pl/hellsoft/slack/wrapper/rtm/SlackRealTimeMessagingClient.kt [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+                type = connect.type
+            } catch (e: Exception) {
+                e { "$e.message" }
+            }
+
+            if ("pong" != type) {
+                i { "Slack RTM message : $text" }
+            }
+
+            if (type != null) {
```
Comment (m0d) [inline slack-api/src/main/java/pl/hellsoft/slack/wrapper/rtm/SlackRealTimeMessagingClient.kt]: this is not needed

Comment (m0d) [inline slack-api/src/main/java/pl/hellsoft/slack/wrapper/rtm/SlackRealTimeMessagingClient.kt]: this check is not necessary

Comment (m0d) [reply slack-api/src/main/java/pl/hellsoft/slack/wrapper/rtm/SlackRealTimeMessagingClient.kt]: im talking about if statement

**Comment by falvick** [comment]
All comments resolved. Please review changes and merge to develop branch

**Pull request merged**